flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1177,10 +1239,8 @@ struct XEmboxOpConversion : public EmboxCommonConversion<fir::XEmboxOp> {
     }
     if (isDerivedType(boxTy))
       TODO("derived type");
-    // Convert descriptor to the prefix type for strong typing.
```

Comment:
I'm not following how this conversion can be omitted now. Why doesn't LLVM IR complain with struct type mismatches?

Reply:
The cast was made wrong in the undefOp + insertOp case since it attemps to cast `alloca` to a memory type, when alloca is actually a `llvm.struct<T>` in the global constant case.

When I analyzed the rational behind the cast, I came to the conclusion that there can actually not be any struct type mismatch here. I think there may have been initially, but `alloca` type is already created with prefix type: https://github.com/flang-compiler/f18-llvm-project/blob/003a9997054c54f8f2aa546187d9a2ecfed533c7/flang/lib/Optimizer/CodeGen/CodeGen.cpp#L1051
So I think this cast is a relicate of when things were different.
I moved the boxTy to be returned by `consDescriptorPrefix` so that we are sure that there is a single point of truth for the type here.

Reply:
Nice.